polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -7,166 +7,367 @@
+                                       AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr = createNamedAttr(
+      createStringAttr(AttrName, Dialect, Ctx), mlir::TypeAttr::get(Ty));
+
+  return Invoke(AddAttrPtr, NamedAttr);
+}
+
+AttrBuilder &AttrBuilder::addAttribute(llvm::Attribute::AttrKind Kind,
+                                       uint64_t Val,
+                                       AddRawIntAttrFuncPtr AddRawIntAttrPtr) {
+  assert(AddRawIntAttrPtr &&
+         "'AddRawIntAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddRawIntAttrFuncPtr AddRawIntAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddRawIntAttrPtr)(Args...);
+  };
+
+  switch (Kind) {
+  case llvm::Attribute::AttrKind::Alignment:
+  case llvm::Attribute::AttrKind::StackAlignment:
+    assert(Val <= llvm::Value::MaximumAlignment && "Alignment too large");
+    return (!Val) ? *this : Invoke(AddRawIntAttrPtr, Kind, Val);
+
+  case llvm::Attribute::AttrKind::Dereferenceable:
+  case llvm::Attribute::AttrKind::DereferenceableOrNull:
+  case llvm::Attribute::AttrKind::UWTable:
+    return (!Val) ? *this : Invoke(AddRawIntAttrPtr, Kind, Val);
+
   default:
     llvm_unreachable("Unexpected attribute kind");
   }
 
   return *this;
 }
 
-AttrBuilder &AttrBuilder::addAttribute(Twine attrName, mlir::Attribute attr) {
-  NamedAttribute namedAttr = createNamedAttr(createStringAttr(attrName), attr);
-  return addAttribute(namedAttr);
+AttrBuilder &AttrBuilder::addAttribute(Twine AttrName, mlir::Attribute Attr,
+                                       AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  NamedAttribute NamedAttr =
+      createNamedAttr(createStringAttr(AttrName, Ctx), Attr);
+  return Invoke(AddAttrPtr, NamedAttr);
 }
 
 AttrBuilder &
-AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind kind) {
-  StringRef attrName = llvm::Attribute::getNameFromAttrKind(kind);
-  return addPassThroughAttribute(createStringAttr(attrName));
+AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind Kind) {
+  return addAttribute(Kind, llvm::None, &AttrBuilder::addPassThroughAttribute);
```

Comment:
Curious, there are different `AttrBuilder::addPassThroughAttribute` that takes different signature, how's does it know which one you are referring here? Does it know by looking into the body of `addAttribute` and find out that it is calling this function pointer with a `NamedAttribute`?